### DISCUSSION TITLE
Possible to move or copy existing element in DOM into a different location

### DISCUSSION THREAD

**Discussion by peace2000** [uBlockO/uBO-Scriptlets#20] [Q&A] [answered]
Hi @uBlock-user ! Does any of your custom scriplets make it possible to copy or move (rather move) a specific element in a DOM tree into a different location in the same DOM tree?

I don't have a specific site I could link.

**Comment by uBlock-user** [comment]
Example ? I can't think of steps to do it unless and untill I have something to work with.

**Comment by peace2000** [comment]
Ok, this is of course totally random example, but:

`https://www.is.fi/`

Move (or copy) element
`.nav-bar__link-bar`
directly under
`#header`
as a first child element.

![kuva](https://user-images.githubusercontent.com/17256841/226670705-7dffa897-0a4c-4258-885a-289509d5e91c.png)

->

![kuva](https://user-images.githubusercontent.com/17256841/226671278-6c2b14f2-2ac3-4d5c-833f-a9097062f65d.png)

**Reply by uBlock-user** [reply]
Use `insert-elem-before` scriptlet of mine where you will need to target `section[class="text-positive"]`.

Alternatively, you can use [**_Element.after()_**](https://developer.mozilla.org/en-US/docs/Web/API/Element/after) method and create a new scriptlet for yourself.

**Reply by peace2000** [reply]
Ok, thanks, can you help with forming the actual rule? Sorry I don't get that example which is listed here https://ubo-scriptlets.pages.dev/scriptlets.js how to form the rule in this particular case.

`is.fi##+js(ieb, .nav-bar__link-bar, section[class="text-positive bg-header-main"])`

Doesn't work, though it's probably just formed in a wrong way.

**Reply by uBlock-user** [reply]
`is.fi##+js(ieb, .nav-bar__link-bar, display:block, section[class="text-positive bg-header-main"], ul)` should work.

You still will have to delete the original by yourself though, my scriptlet doesn't do it.

**Reply by peace2000** [reply]
For some reason, it doesn't work. Tested on FF and on Chrome.

**Reply by uBlock-user** [reply]
Yeah seems so. Don't know why.

**Reply by peace2000** [reply]
I made some tests with another site.

https://www.ampparit.com/

I'm trying to copy element `.social-media-feedback-link` before `#react-view`.

This one seems to work fine:

`ampparit.com##+js(ieb, .social-media-feedback-link, , [id="react-view"], div)`

But this doesn't:

`ampparit.com##+js(ieb, [class="social-media-feedback-link"], , [id="react-view"], div)`

For me it seems that square bracket form doesn't work in the element-to-be-copied. Is this intentional?

**Reply by uBlock-user** [reply]
>  Is this intentional?

No. Why would it be ?

**Comment by uBlock-user** [comment]
> [class="social-media-feedback-link"]

Must be slow to evaluate in the browser itself. My scriptlet has no flaws, so the issue is not on my side.

Try this version -- 


```
/// iebt.js
(() => {
		const identifier = '{{1}}';
		if ( identifier === '' || identifier === '{{1}}' ) { return; }
		const identifiers = identifier.split(/\s*\|\s*/);
		const behavior = '{{5}}';
		const insertelem = () => {
						let timer = undefined;
						try {
							for (const token of identifiers) {
							     const element = document.createElement('{{4}}');
							     const node = document.querySelector('{{3}}');
							     if (token.charAt(0) === '#') {
							    	 element.id = token.substring(1);
							     } else if (token.charAt(0) === '.') {
							    	 element.className = token.substring(1);
							     } else { return; }	 
							     element.style.cssText = '{{2}}';
							     document.body.insertBefore(element, node);
							}	
						} catch { }
	   	};
		const mutationHandler = mutations => {
			if ( timer !== undefined ) { return; }
			let skip = true;
			for ( let i = 0; i < mutations.length && skip; i++ ) {
			    const { type, addedNodes, removedNodes } = mutations[i];
			    if ( type === 'attributes' ) { skip = false; }
			    for ( let j = 0; j < addedNodes.length && skip; j++ ) {
				if ( addedNodes[j].nodeType === 1 ) { skip = false; break; }
			    }
			    for ( let j = 0; j < removedNodes.length && skip; j++ ) {
				if ( removedNodes[j].nodeType === 1 ) { skip = false; break; }
			    }
			}
			if ( skip ) { return; }
			timer = self.requestIdleCallback(insertelem, { timeout: 10 });
		    };
		    const start = ( ) => {
			insertelem();
			if ( /\bloop\b/.test(behavior) === false ) { return; }
			const observer = new MutationObserver(mutationHandler);
			observer.observe(document.documentElement, {
			    attributes: true,
			    childList: true,
			    subtree: true,
			});
		    };
		    if ( document.readyState !== 'complete' && /\bcomplete\b/.test(behavior) ) {
			self.addEventListener('load', start, { once: true });
		    } else if ( document.readyState === 'loading' ) {
			self.addEventListener('DOMContentLoaded', start, { once: true });
		    } else {
			start();
		    }
})();
```

Filter `is.fi##+js(iebt, .nav-bar__link-bar, display:block, section[class="text-positive bg-header-main"], ul, loop)`

**Reply by peace2000** [reply]
`is.fi##+js(iebt, .nav-bar__link-bar, display:block, section[class="text-positive bg-header-main"], ul, loop)`

Doesn't work either. Tested with FF and Chrome.

Also I got the same results with these as I got with `ieb`:

`ampparit.com##+js(iebt, .social-media-feedback-link, , [id="react-view"], div, loop)` - works

`ampparit.com##+js(iebt, [class="social-media-feedback-link"], , [id="react-view"], div, loop)` - doesn't

**Reply by uBlock-user** [reply]
Try `is.fi##+js(iebt, .nav-bar__link-bar, , section[class="text-positive bg-header-main"], ul, loop)`

**Reply by peace2000** [reply]
Nope, still not... Does that work on your end?

**Reply by uBlock-user** [reply]
I think, I misunderstood your question. You want to place that element as first-child. For that you will need to use [**_Element.prepend()_**](https://developer.mozilla.org/en-US/docs/Web/API/Element/prepend)

**Reply by uBlock-user** [reply]
![Capture](https://user-images.githubusercontent.com/21290713/227156367-87951688-03fb-4316-8608-5c646a86ae5e.JPG)

@peace2000 Is that what you want ?

**Reply by peace2000** [reply]
In this case, yes but not necessarily universally. The point is to be able to place element X before element Y (element X is not always supposed to be the first child, it just needs to be before element Y).

Sorry if I'm unclear.

**Reply by uBlock-user** [reply]
> element X is not always supposed to be the first child,

[**_Element.prepend()_**](https://developer.mozilla.org/en-US/docs/Web/API/Element/prepend) is specific for inserting as `first-child` only. Will need to at `element.before` and `element.after` then.

**Comment by uBlock-user** [comment, marked as answer]
> The point is to be able to place element X before element Y

```
/// insert-child-before.js
// is.fi##+js(insert-child-before, section.text-positive.bg-header-main, .nav-bar__link-bar)
 (() => {
		const selector = '{{1}}';
		if ( selector === '' || selector === '{{1}}' ) { return; }
	        const element = '{{2}}';
		if ( element === '' || element === '{{2}}' ) { return; }
		const behavior = '{{3}}';
		let timer = undefined;
		const insertelem = () => {
			try {
				const elems = document.querySelectorAll(selector);
				const nodes = document.querySelectorAll(element);
				for (let i = 0; i < elems.length; i++) {
				    elems[i].before(nodes[i]);
				}	
			} catch { }
	   	};
		const mutationHandler = mutations => {
			if ( timer !== undefined ) { return; }
			let skip = true;
			for ( let i = 0; i < mutations.length && skip; i++ ) {
			    const { type, addedNodes, removedNodes } = mutations[i];
			    if ( type === 'attributes' ) { skip = false; }
			    for ( let j = 0; j < addedNodes.length && skip; j++ ) {
				if ( addedNodes[j].nodeType === 1 ) { skip = false; break; }
			    }
			    for ( let j = 0; j < removedNodes.length && skip; j++ ) {
				if ( removedNodes[j].nodeType === 1 ) { skip = false; break; }
			    }
			}
			if ( skip ) { return; }
			timer = self.requestIdleCallback(insertelem, { timeout: 10 });
		};
		const start = ( ) => {
			insertelem();
			if ( /\bloop\b/.test(behavior) === false ) { return; }
			const observer = new MutationObserver(mutationHandler);
			observer.observe(document.documentElement, {
			    attributes: true,
			    childList: true,
			    subtree: true,
			});
		};
		if ( document.readyState !== 'complete' && /\bcomplete\b/.test(behavior) ) {
		     self.addEventListener('load', start, { once: true });
		    } else if ( document.readyState === 'loading' ) {
		     self.addEventListener('DOMContentLoaded', start, { once: true });
		    } else {
		     start();
		}
})();
```

@peace2000 This works for inserting an element before any child element of a parent element.

**Reply by peace2000** [reply]
@uBlock-user Hi, thanks your icb and ica scripts work fine!

Though `is.fi` seems to fight back (using `loop` is required) other sites I tested with are ok. Great!

PS. You might want to move this discussion into your scriplet repo, this was wrong repo I accidentally chose.

**Reply by uBlock-user** [reply]
> using loop is required)

Should work with `complete` too. On my side it works with the above filter.
